Add timeline

quinnyyy opened the pull request:
first attempt

resolves #15

quinnyyy committed:
Add timeline functionality and fix broken tests

quinnyyy committed:
Add tests for parseDate

quinnyyy committed:
Update toString to format dates

quinnyyy committed:
Update textui test

Zhilin-Huang commented on src/main/java/seedu/tp/flashcard/FlashcardList.java:
```diff
@@ -68,6 +68,15 @@ public int getTotalFlashcardNum() {
+     * 
+     * @return the list of flashcards
+     */
+    public ArrayList<Flashcard> getFlashcards() {
```
Consider using `List<Flashcard>` interface instead of the `ArrayList<Flashcard>` class

Zhilin-Huang commented on src/main/java/seedu/tp/ui/Ui.java:
```diff
@@ -75,6 +80,55 @@ public String promptUserForRequiredField(String fieldName) {
+     * @param fieldName string representing name of the date to prompt for
+     * @return the parsed date
+     */
+    public Optional<LocalDate> promptUserForOptionalLocalDate(String fieldName) {
```
If the date is optional, why do we still prompt user to enter again when the input is empty?

quinnyyy replied:
When the input is empty it returns Optional.empty().
When the input is non-empty is tries to parse a date. If the date isn't parseable then it will ask again

Zhilin-Huang replied:
Understood. Sorry I missed the `EMPTY_STRING` part!

Zhilin-Huang commented on src/main/java/seedu/tp/ui/Ui.java:
```diff
@@ -103,6 +157,25 @@ public void listAllFlashcards(FlashcardList flashcardList) {
+            return;
+        }
+        
+        ArrayList<Flashcard> flashcards = new ArrayList<>(flashcardList.getFlashcards());
```
Use `List<>` instead of `ArrayList` when defining the data type

Zhilin-Huang reviewed (changes requested):
I suspect that we need to resolve the merge conflicts before the CI tests can run
Also consider adding JUnit tests for `TimelineCommand`

quinnyyy committed:
Use List interface instead of ArrayList

quinnyyy committed:
Merge branch 'master' of https://github.com/AY1920S2-CS2113-T14-1/tp into Add-Timeline

# Conflicts:
#	src/main/java/seedu/tp/parser/Parser.java
#	src/main/java/seedu/tp/ui/Ui.java
#	src/main/java/seedu/tp/utils/Constants.java

quinnyyy committed:
Fix merge conflict I forgot about

quinnyyy committed:
Add TimelineCommandTest

quinnyyy committed:
Remove unused import

quinnyyy commented:
I resolved the merge conflicts and added TimelineCommand test. 
I changed the ArrayLists to List interfaces but a little confused about your second comment @Zhilin-Huang

Zhilin-Huang commented on src/test/java/seedu/tp/commands/TimelineCommandTest.java:
```diff
@@ -0,0 +1,58 @@
+    private final PrintStream backupStdout = System.out;
+    
+    @BeforeEach
+    public void captureStdout() {
```
Nice! I'll use this in other JUnit tests

Zhilin-Huang commented:
> I resolved the merge conflicts and added TimelineCommand test.
> I changed the ArrayLists to List interfaces but a little confused about your second comment @Zhilin-Huang

Thanks for explaining to me about the second comment :) I think this PR looks good to merge
